### ISSUE TITLE
请问action是一个list，如何修改DQN的agent？

### ISSUE THREAD

**Issue by yeyupiaoling** [PaddlePaddle/PARL#434]
例如游戏的action是这样的`[1 0 0 0 1 0 1 0 1]`, 我应该修改agent和model？

**Comment by TomorrowIsAnOtherDay** [comment]
每个位置的取值都是0或者1吗？如果是这样的话，建议让DQN直接输出2^9次方（512）个action，然后每个action再用二进制表示成对应的action。

**Comment by yeyupiaoling** [comment]
@TomorrowIsAnOtherDay 谢谢回答。但这样模型的输出层是不是有点大了呢？

**Comment by TomorrowIsAnOtherDay** [comment]
是的，一般我们不会在DQN里面使用那么多的输出，建议重新对问题建模。
或者尝试下我们的进化算法，我们最近在这种离散空间大的问题上使用ES取得了不错的效果。

**Comment by yeyupiaoling** [comment]
> 每个位置的取值都是0或者1吗？如果是这样的话，建议让DQN直接输出2^9次方（512）个action，然后每个action再用二进制表示成对应的action。

@TomorrowIsAnOtherDay 它慢慢预测趋向一个值是怎么回事呢？

**Comment by yeyupiaoling** [comment]
@TomorrowIsAnOtherDay 你说我sigmoid作为激活函数这样合不合理呢？这个也是，预测都慢慢趋向一个组合动作
model
```python
self.fc3 = layers.fc(size=9, act='sigmoid')
```

agent

```python
    def predict(self, obs):
        obs = np.expand_dims(obs, axis=0)
        pred_Q = self.fluid_executor.run(program=self.pred_program,
                                         feed={'obs': obs.astype('float32')},
                                         fetch_list=[self.value])[0][0]
        act = [int(a + 0.5) for a in pred_Q]
        return act
```

**Comment by TomorrowIsAnOtherDay** [comment]
我们在issue区主要是讨论框架的相关使用问题哈，请与你身边的同学讨论RL的应用问题。

**Comment by ouyangshixiong** [comment]
PARL内部已经帮你做了one-hot,不需要自己搞

>     def learn(self,
              obs,
              action,
              reward,
              next_obs,
              terminal,
              learning_rate=None):
        """ update value model self.model with DQN algorithm
        """
        # Support the modification of learning_rate
        if learning_rate is None:
            assert isinstance(
                self.lr,
                float), "Please set the learning rate of DQN in initializaion."
            learning_rate = self.lr

        pred_value = self.model.value(obs)
        next_pred_value = self.target_model.value(next_obs)
        best_v = layers.reduce_max(next_pred_value, dim=1)
        best_v.stop_gradient = True
        target = reward + (
            1.0 - layers.cast(terminal, dtype='float32')) * self.gamma * best_v

        action_onehot = layers.one_hot(action, self.act_dim)
        action_onehot = layers.cast(action_onehot, dtype='float32')
        pred_action_value = layers.reduce_sum(
            layers.elementwise_mul(action_onehot, pred_value), dim=1)
        cost = layers.square_error_cost(pred_action_value, target)
        cost = layers.reduce_mean(cost)
        optimizer = fluid.optimizer.Adam(
            learning_rate=learning_rate, epsilon=1e-3)
        optimizer.minimize(cost)
        return cost

**Comment by rical730** [comment]
> PARL内部已经帮你做了one-hot,不需要自己搞

非常欢迎参与PARL的issue讨论，训练内部的action需要转换的地方PARL已经封装好了，但提问者提问的应该是predict得到的action，而不是learn中途的action
